Sequential testing

Eforen opened the pull request:
## What it does
* Adds a fixture for Log Levels
* Adds `<seq>`, `<set>`, and `<iter>` special headers that define the sequence to run things in the set it belongs to and the max iterations for it to call propagate.
* Adds a `<DC>` value that is case insensitive that makes the test vector ignore that value so that we don't have to care about it in some cases where we do care about it in others.
* Adds a `<FLOAT>` value that is also case insensitive that only is true if the value is floating. (This is basicly the same as the UUUUU values previously but is more clear.
* Adds many tests for the test vectors behavior and formatting
* Adds more helpful error messages to the too many bits message so that it is helpful instead of just annoying.
* Adds support for optional ignored _ chars for visual spacing in test values

## This is for issues:
* Test vectors for sequential circuits #947
* Revamp the test vector component to add better support for sequential circuits testing #598
* Test vector from command line documentation #830
* Create a library of test circuits #45 ( I have a DLatch test in my code that can be expanded to cover other tests.

## May also cover issues:
* Cannot use "don't care" character "x" on inputs of the vector test function #1123
* Test Vectors - Variables #399
* Problem with the test vectors. #193
* Command line option -test does not seem to work on 2.14.6 #185
  * I used the command line options a lot developing this PR (I must have run like 200 times)

davidhutchens commented:
I have found the main problem with the example not working in your PR. It is related to things I'm working on with issue #2404. You are, in fact, not running with the simulation thread! You can fix the issue for testing purposes by moving `moveNonSimThreadEvents();` in Propagator.propagate() from the top of the method to just before the while loop.

Specifically, because you are not using the simulation thread, the processing of the dirty components is putting the simulation events in the bypass list and they are not in the queue for the loop.

The full fix should be with the PR for #2404 that I am working on.

Eforen commented:
I was a little confused by the documentation stuff too. I'll work on that I just wanted to get this PR up to make sure y'all are receptive to it before spending too much time working on docs. 

I'll sort out those linter issues soon too.

Eforen commented:
I have cleaned up the internal documentation and added translations outside of English as much as possible. (The translations were via large amounts of help via AI I don't know these languages so they are as good as I can do but the look correct and seem to translate back to roughly the same info)

Eforen commented:
I believe I can now say it is ready.

davidhutchens commented on src/main/java/com/cburch/logisim/circuit/Circuit.java:
```diff
@@ -464,24 +464,71 @@ public boolean doTestBench(Project project, Instance[] pin, Value[] val) {
+    // We loop until it returns false (stable) or we hit max iterations
     try {
-      prop.propagate();
+      for (int iter = 0; iter < iterations; iter++) {
```
I made PR #2415 to fix the issue in the propagator that caused it to sometimes faillto fully propagate. You can merge with that if you want to try removing that loop.

davidhutchens reviewed (changes requested):
The code looks good. I still think the iter loop is not needed in Circuit. I made PR #2415 to fix the issue in the Propagator. You can include that to test if you wish.

I noted a problem in a table in the documentation. That should be fixed in all languages. Otherwise, the documentation looks good to me. @maehne should have a look at the documentation as well.

Other than that, this PR looks good to me.

Eforen commented on src/main/resources/doc/en/html/guide/test_vector/test_vector_sequential.html:
```diff
@@ -0,0 +1,101 @@
+<li>The first three tests are combinational (seq=0), so the circuit resets
+between each test, even though they share set 0.</li>
+<li>The next three tests all have set 1 with seq values 1-3, so they form one sequence
+that runs in order without reset.</li>
```
The Set is the group of tests (Seperation) the seq is the order to run a set in if they are not seq=0. Say for example you have to set a to 1 and then 2 and then 1 again, but you also want to check 3 then 2 then 3 but you can't have it go from 1 to 3. and you also want to check some one offs like a=4 and a=5 but don't want to keep the state between them. In this case you can do this
| set| seq| a|
|--:|:--|:-:|
|1|1|1|
|1|2|2|
|1|3|1|
|2|1|3|
|2|2|2|
|2|3|3|
|0|0|4|
|0|1|5|

This will do the following in this order:
1. Reset the circuit to fresh state
2. set a=4
3. Reset the circuit to fresh state
4. set a=5
5. Reset the circuit to fresh state
6. set a=1
7. set a=2
8. set a=1
9. Reset the circuit to fresh state
10. set a=3
11. set a=2
12. set a=3

Now obviously you would want to actually have something be tested here but I just wanted to show the order of execution stuff not a full test suite.

davidhutchens replied:
Here is a screenshot from the 12 Dec. 2025 version. It is at the bottom of the Sequential Testing section.

<img width="695" height="420" alt="Example" src="https://github.com/user-attachments/assets/24632b10-75c2-4c5a-a1c8-19566f5121b5" />

As you can see, all the Set values in this example are 0. But the description says the first three are 0 and the next three are 1. The description does not match the example. This is what this conversation referred to. I suggest fixing the example to match the description.

Eforen replied:
ooops I miss understood your previous >.<I totally agree we should fix that I will do it right now.

davidhutchens replied:
Your update looks good. I'll close this conversation as resolved.

Eforen commented:
I think the academics are gona love this change!

davidhutchens commented on src/main/java/com/cburch/logisim/gui/test/TestPanel.java:
```diff
@@ -300,18 +296,20 @@ public void modelChanged(Model oldModel, Model newModel) {
+      activeShowRows.clear();
+      activeSetRows.clear();
+      storedPinValues = null;
+      table.setModel(newModel == null ? null : this);
```
Lines 299-302 in TestPanel are indented 2 spaces too far.

davidhutchens commented on src/main/resources/doc/en/html/guide/test_vector/test_vector_sequential.html:
Iter is still discussed in lines 33-38 in test_vector_sequential.html. And it appears as a column in the following table, lines 43-51.

Please remove that. Here is a picture:

<img width="693" height="568" alt="Image" src="https://github.com/user-attachments/assets/5c6e3ed1-7dde-40e8-9c56-a1bca81f9ea1" />

Eforen replied:
Thought I took all that stuff our sorry about that. Documentation while important sometimes makes ones eyes glaze over when looking at it too long.

Eforen replied:
It's just some deletes I understand the languages in question enough to make these edits.

davidhutchens reviewed (changes requested):
I noted two minor linter problems in the code that should be fixed. Otherwise, I think the code is ready.

I note that there were race conditions preexisting in the TestVector code. So I will be refactoring this code to eliminate those. Fixing those race conditions is beyond the scope of this PR and merging it will help me fix the issues as a group.

I also noted that `Iter` still exists in the English documentation on the Sequential Testing page. Please remove that. Once that is removed, you should update the documentation in the other languages. It is quite acceptable to simply copy the English pages to the other languages and let those that are fluent in those languages write the translation.

With those minor changes, this PR should be good.

Eforen commented on src/main/java/com/cburch/logisim/gui/test/TestPanel.java:
```diff
@@ -376,105 +350,194 @@ public void handleButtonClick(int displayRow) {
+      
+      if (prop.isOscillating()) {
+        throw new TestException("Oscillation detected at sequence step " + 
+            (vec.seqNumbers != null && stepRow < vec.seqNumbers.length ? vec.seqNumbers[stepRow] : 0));
```
Makes sense Will Fix Soon

maehne commented on docs/test_vector.md:
What additional information does this file provide compared to the documentation integrated into the User's Guide? To make it more visible, consider to reference it in the project's `README.md` including a "[« Go back](https://github.com/logisim-evolution/logisim-evolution/blob/main/README.md)" link as the other MarkDown files in the `docs/` directory do.
